pubsub/streaming-analytics/PubSubToGCS_test.py

```diff
@@ -63,19 +64,28 @@ def test_run(publisher_client, topic_path):
 
     # Use one process to publish messages to a topic.
     publish_process = mp.Process(
-        target=lambda: _infinite_publish_job(publisher_client, topic_path))
+        target=lambda: _infinite_publish_job(publisher_client, topic_path)
+    )
 
     # Use another process to run the streaming pipeline that should write one
     # file to GCS every minute (according to the default window size).
     pipeline_process = mp.Process(
-        target=lambda: sp.call([
-            'python', 'PubSubToGCS.py',
-            '--project', PROJECT,
-            '--runner', 'DirectRunner',
-            '--temp_location', tempfile.mkdtemp(),
-            '--input_topic', topic_path,
-            '--output_path', 'gs://{}/pubsub/{}/output'.format(BUCKET, UUID),
-        ])
+        target=lambda: sp.call(
```

Comment:
This may fail because you can't be certain of the DAC on the subprocess, maybe pass the path to the service account credential as well and use this to authenticate the client.

Reply:
Hmm, it's failing locally too. What is meant by DAC?

Reply:
Oh, sorry! DAC is default application credentials, set by running `gcloud auth application-default login`. I'm just saying that the sample or gsutil or gcloud may not be authorizing correctly.

Reply:
I should have said ADC, not DAC, very confusing. Application default credentials.

Reply:
Dataflow's DirectRunner shouldn't need extra permissions. And this bucket `BUCKET = os.environ["CLOUD_STORAGE_BUCKET"]` is a public bucket. 

When I tried `pytest PubSubToGCS_test.py` and the test passed.
However, `nox -s "py36(sample='./pubsub/streaming-analytics')"` failed. 

@tswast or @gguuss any idea as to why?

Reply:
@anguillanneuf  Maybe something specific to the version of python that Nox is  testing? What version of Python are you using with pytest?

Reply:
I'm seeing:

```
__________________________________ test_run ___________________________________
Traceback (most recent call last):
  File "/tmpfs/src/github/python-docs-samples/pubsub/streaming-analytics/PubSubToGCS_test.py", line 112, in test_run
    assert len(files) > 0
AssertionError: assert 0 > 0
 +  where 0 = len({})
----------------------------- Captured stderr call -----------------------------
INFO:root:Running pipeline with DirectRunner.
```

The code surrounding the test is:
```
    # Check for output files on GCS.
    gcs_client = beam.io.gcp.gcsio.GcsIO()
    # This returns a dictionary.
    files = gcs_client.list_prefix('gs://{}/pubsub/{}'.format(BUCKET, UUID))
    assert len(files) > 0
```

Has the output from `beam.io.gcp.gcsio` changed?

Reply:
When I step through the code locally, I'm seeing the GCS object in beam try to auth:

```
  "error": {
    "code": 401,
    "message": "Anonymous caller does not have storage.objects.list access to gs://python-test-bucket/.",
    "errors": [
      {
        "message": "Anonymous caller does not have storage.objects.list access to python-test-bucket",
        "domain": "global",
        "reason": "required",
        "locationType": "header",
        "location": "Authorization"
      }
    ]
  }
}
```

When I am able to authenticate, I'm seeing the call to `len(gcs_client.list_prefix(...))` is working so it's not that.

Reply:
I used pytest with python 3.6.4, which is identically to what `nox` uses. 

![Screen Shot 2019-12-09 at 1 22 01 PM](https://user-images.githubusercontent.com/9518298/70474182-fe250400-1a86-11ea-87cd-342ce059cc9e.png)

Reply:
I suspect that the DirectRunner from all previous jobs are not stopped properly. Every time I run a test, the Dataflow job continues to be active when I list running processes on my machine `ps aux`. I have to manually kill by PID.

Reply:
@crwilcox As a `nox` maintainer, do you have any guesses as to why `pytest` and `nox` behave differently for [`PubSubToGCS_test.py`](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/pubsub/streaming-analytics/PubSubToGCS_test.py)?

Reply:
I just want to comment publicly admiring your detective work thus far and saying thank you for such good documentation of it here 🔍 👏

Reply:
@GoogleCloudPlatform/python-samples-owners Is there any chance that these processes could be on the loose wherever kokoro runs? I actually don't even know where to look and if someone is going digging, would love to be present and watch the steps you take.

Reply:
@anguillanneuf  Nox will create a new virtual environment and install the latest version of the dependencies for each test run. If you haven't already done this, you can try creating an empty virtualenv, installing the test dependencies, and then running pytest.

Reply:
I think I've figured out why 😉 Hope to finally get this done tomorrow.